api/v1alpha1/adminnetworkpolicy_types.go

```diff
@@ -0,0 +1,309 @@
+type AdminNetworkPolicyPeer struct {
+	// Namespaces defines a way to select a set of Namespaces
+	// +optional
+	Namespaces *NamespaceSet `json:"namespaces,omitempty"`
+	// NamespacedPods defines a way to select a specific set of pods in
+	// in a set of namespaces
+	// +optional
+	NamespacedPods *NamespaceAndPodSet `json:"namespacedPods,omitempty"`
+}
+
+// NamespaceSet defines a flexible way to select Namespaces in a cluster.
+// Exactly one of the selectors should be set.  If a consumer observes none of
+// its fields are set, they should assume an unknown option has been specified
+// and fail closed.
+type NamespaceSet struct {
+	// Self cannot be "false" when it is set.
+	// If Self is "true" then all pods in the subject's namespace are selected.
+	// +optional
+	Self *bool `json:"self,omitempty"`
+
+	// NotSelf cannot be "false" when it is set.
+	// if NotSelf is "true" then all pods not in the subject's Namespace are selected.
+	// +optional
+	NotSelf *bool `json:"notSelf,omitempty"`
+
+	// NamespaceSelector is a labelSelector used to select Namespaces, This field
+	// follows standard label selector semantics; if present but empty, it selects
+	// all Namespaces.
+	// +optional
+	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector,omitempty"`
+
+	// SameLabels is used to select a set of Namespaces that share the same values
+	// for a set of labels.
+	// To be selected a Namespace must have all of the labels defined in SameLabels,
+	// and they must all have the same value as the subject of this policy.
+	// If Samelabels is Empty then nothing is selected.
+	// +optional
+	SameLabels []string `json:"sameLabels,omitempty"`
+
+	// NotSameLabels is used to select a set of Namespaces that do not have a set
+	// of label(s). To be selected a Namespace must have none of the labels defined
+	// in NotSameLabels. If NotSameLabels is empty then nothing is selected.
+	// +optional
+	NotSameLabels []string `json:"notSameLabels,omitempty"`
+}
+
+// PodSet defines a flexible way to select pods in a cluster. Exactly one of the
+// selectors should be set.  If a consumer observes none of its fields are set,
+// they should assume an unknown option has been specified and fail closed.
+type PodSet struct {
+	// PodSelector is a labelSelector used to select Pods, This field is NOT optional,
+	// follows standard label selector semantics and if present but empty, it selects
+	// all Pods.
+	PodSelector *metav1.LabelSelector `json:"podselector"`
+}
+
+// NamespaceSetAndPod defines a flexible way to select Namespaces and pods in a
+// cluster. The `Namespaces` and `PodSelector` fields are required.
+type NamespaceAndPodSet struct {
+	// Namespaces is used to select a set of Namespaces.  It must be defined and
+	// non-empty.
+	Namespaces NamespaceSet `json:"namespaces"`
+
+	// Pods is used to select a set of Pods in the set of Namespaces. It must
+	// must be defined and non-empty.
+	Pods PodSet `json:"pods"`
```

Comment:
saying it must be non-empty is confusing given that it only contains a single field which itself may be empty. Though also, it seems like it would make sense to allow omitting `Pods` if you wanted to select all pods.

We really shouldn't be encouraging admins to define ANPs down to the Pod level, given that admins don't control the pod labels in most namespaces. Saying that only pods labeled `secure: totally!` can access Thing X is not useful when everyone can just add that label to their own pods.

Reply:
Right.... This is one of those that is funny in order to provide the ability to one day provide extra selectors. 

> Though also, it seems like it would make sense to allow omitting Pods if you wanted to select all pods. 

Agreed but then to keep things consistent we would need to do the same thing for `Namespaces`..... I think although it's a little clunky 
```
pods: 
   podSelector: {}
```

Isn't a horrible way to express "all pods" 

> We really shouldn't be encouraging admins to define ANPs down to the Pod level, given that admins don't control the pod labels in most namespaces. Saying that only pods labeled secure: totally! can access Thing X is not useful when everyone can just add that label to their own pods.

Agreed, however they still should to able to express such desires... especially when concerning pods in system/ other admin controller namespaces.  Generally I think this sediment should be combated with really good "best practices" documentation.